python/lsst/meas/algorithms/dynamicDetection.py

```diff
@@ -175,12 +182,38 @@ def calculateThreshold(self, exposure, seed, sigma=None):
                 & np.isfinite(fluxes) & np.isfinite(area) & np.isfinite(bg))
 
         minNumSources = int(self.config.minFractionSources*self.skyObjects.config.nSources)
+        # Reduce the number of sky sources required if requested, but ensure
+        # a minumum of 3.
+        if minFractionSourcesFactor != 1.0:
+            minNumSources = max(3, int(minNumSources*minFractionSourcesFactor))
         if good.sum() < minNumSources:
-            raise NoWorkFound(f"Insufficient good sky source flux measurements ({good.sum()} < "
-                              f"{minNumSources}) for dynamic threshold calculation.")
-
-        self.log.info("Number of good sky sources used for dynamic detection: %d (of %d requested).",
-                      good.sum(), self.skyObjects.config.nSources)
+            if not isBgTweak:
+                msg = (f"Insufficient good sky source flux measurements ({good.sum()} < "
+                       f"{minNumSources}) for dynamic threshold calculation.")
+            else:
+                msg = (f"Insufficient good sky source flux measurements ({good.sum()} < "
+                       f"{minNumSources}) for background tweak calculation.")
+
+            nPix = exposure.mask.array.size
+            badPixelMask = lsst.afw.image.Mask.getPlaneBitMask(["NO_DATA", "BAD"])
+            nGoodPix = np.sum(exposure.mask.array & badPixelMask == 0)
+            if nGoodPix/nPix > 0.2:
+                detectedPixelMask = lsst.afw.image.Mask.getPlaneBitMask(["DETECTED", "DETECTED_NEGATIVE"])
+                nDetectedPix = np.sum(exposure.mask.array & detectedPixelMask != 0)
+                msg += (" However, there are Number (%) of good data pixels (i.e. not NO_DATA or BAD): "
+                        f"{nGoodPix} ({100*nGoodPix/nPix:.1f}%) in this patch, so there should be "
+                        "sufficient area to locate suitable sky sources.  Something is amiss. "
+                        "(Note that the fraction of \"good\" pixels that have the DETECTED or "
+                        f"DETECTED_NEGATIVE mask bit set is {100*nDetectedPix/nGoodPix:.1f}%).")
```

Comment:
Propose rewording this:

> However, {nGoodPix}/{nPix} pixels are not marked NO_DATA or BAD, so there should be sufficient area to locate suitable sky sources.  Note that {nDetectedPix} were marked as DETECTED or DETECTED_NEGATIVE.

It was the "However there are Number (%s)" beginning in particular that didn't read well to me, but I also have a slight preference for putting the original numbers into the logs directly, even if that makes the reader compute some percentages themself.